assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js

```diff
@@ -0,0 +1,196 @@
+/**
+ * SettingsEnhancedMeasurementSwitch component.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
+import {
+	ENHANCED_MEASUREMENT_ENABLED,
+	ENHANCED_MEASUREMENT_FORM,
+	MODULES_ANALYTICS_4,
+	PROPERTY_CREATE,
+	WEBDATASTREAM_CREATE,
+} from '../../datastore/constants';
+import EnhancedMeasurementSwitch from '../common/EnhancedMeasurementSwitch';
+import {
+	isValidPropertyID,
+	isValidPropertySelection,
+	isValidWebDataStreamID,
+	isValidWebDataStreamSelection,
+} from '../../utils/validation';
+const { useSelect, useDispatch } = Data;
+
+export default function SettingsEnhancedMeasurementSwitch( {
+	hasAnalytics4Access,
+} ) {
+	const isEnhancedMeasurementEnabled = useSelect( ( select ) =>
+		select( CORE_FORMS ).getValue(
+			ENHANCED_MEASUREMENT_FORM,
+			ENHANCED_MEASUREMENT_ENABLED
+		)
+	);
+
+	const propertyID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getPropertyID()
+	);
+
+	const webDataStreamID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
+	);
+
+	const isLoadingProperties = useSelect( ( select ) => {
+		return select( MODULES_ANALYTICS_4 ).isLoadingProperties( {
+			hasModuleAccess: hasAnalytics4Access,
+		} );
+	} );
+
+	const isLoadingWebDataStreams = useSelect( ( select ) => {
+		return select( MODULES_ANALYTICS_4 ).isLoadingWebDataStreams( {
+			hasModuleAccess: hasAnalytics4Access,
+		} );
+	} );
+
+	const isEnhancedMeasurementStreamEnabled = useSelect( ( select ) => {
+		if (
+			! isValidPropertyID( propertyID ) ||
+			! isValidWebDataStreamID( webDataStreamID ) ||
+			isLoadingProperties ||
+			isLoadingWebDataStreams
+		) {
+			return null;
```

Comment:
It seems a bit odd to return `null` here given this is essentially a loading state, which we tend to return `undefined` for instead. 🤔

Reply:
That's a good point - I've updated to return `undefined` here for the loading case. Please note, though, I've kept the `null` for the non-valid `propertyID` or `webDataStreamID` case, because this doesn't necessarily indicate a loading state per se, they can be in the `PROPERTY_CREATE` or `WEBDATASTREAM_CREATE` states and we don't want to return `undefined` in this case as `isEnhancedMeasurementStreamEnabled` being `undefined` value results in the `useEffect()` below being skipped - and we want it to run when `PROPERTY_CREATE` or `WEBDATASTREAM_CREATE` is selected.